design/control-data-plane-separation/design.md

```diff
@@ -0,0 +1,770 @@
+this option as a last resort that we can decide to adopt if we hit a roadblock with the nginx agent.
+
+[controller]: https://docs.nginx.com/nginx-controller/api/overview/
+
+[nplus]: http://nginx.org/en/docs/http/ngx_http_api_module.html
+
+[xds]: https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol
+
+[xds-wg]: https://github.com/cncf/xds
+
+## Deployment Architecture
+
+![Deployment architecture](deployment-architecture.png)
+
+* _Control Plane Deployment_: The control plane is a Kubernetes Deployment with one container running the NKG
+  controller. Initially, the control plane will be limited to a single Pod. Once we add leader election, the control
+  plane will be able to scale. The control plane will perform the same functions as it does today, but instead of
+  configuring nginx by writing files to a shared volume, it will send the configuration to the agent via gRPC.
+* _Control Plane Service_: Exposes the control plane via a Kubernetes Service of type `ClusterIP`. The data plane will
+  use the DNS name of the Service to connect to the control plane.
+* _Data Plane DaemonSet/Deployment_: A user can deploy the data plane as either a DaemonSet or Deployment. The data
+  plane contains a single container running both the agent and nginx processes. The agent will download the
+  configuration from the control plane over a streaming RPC.
+* _NGINX Service_: Exposes nginx via a Kubernetes Service of type `LoadBalancer .`This is the entry point for the
+  customer’s traffic. Initially, this Service will only expose ports 80 and 443. In the future, if we add support for
+  additional listener ports, this Service will expose all the listener ports. Note that this Service should not expose
+  any of the agent’s ports.
+
+The user will deploy the control plane, and data plane via Kubernetes manifests. The control plane supports a
+single `GatewayClass` and  `Gateway` resource. Following this model, it makes sense for the control plane to support a
+single data plane Deployment/DaemonSet. Each agent+nginx Pod in the data plane will receive the same configuration from
+the control plane. In the future, we may reconsider this model if there’s a use case for the control plane to support
+multiple `Gateway` resources. For example, in that model, it might make sense for the control plane to be responsible
+for provisioning a separate data plane for each `Gateway.`
+
+### Running Multiple NKG Controllers
+
+Users can run multiple NKG controllers in the same cluster by leveraging the `GatewayClass` resource. Multiple NKG
+controllers will be able to coexist in the same cluster as long as they each have a unique corresponding `GatewayClass`
+resource. In this case, each installation of NKG will contain a unique `GatewayClass` resource, a control plane
+Deployment, and a data plane Deployment/DaemonSet.
+
+## Communication Channels
```

Comment:
How resilient is the agent in the face of communication channels problems?

For example:
- failures to establish connections with the control plane (fast and slow )
- termination of existing connections. for example, after the control plane asked  the data plane to download the config, but before the downloading has finished.

Reply:
I suggest we make sure to cover those cases in our testing for the implementation PRs